### DISCUSSION TITLE
pynetbox importing modules and adopt_compnents

### DISCUSSION THREAD

**Discussion by wannagofast83** [netbox-community/netbox#18028] [Help Wanted!]
I have tried a number of ways to get a script to import modules and then adopt components and it fails every time because there is already a interface with that name on the device. 

`nb.dcim.modules.create(
                device = nb_device.id,
                module_bay = nb_device_module_bay.id,
                module_type = nb_module_type.id,
                status = "active",
                serial = _module_serial,
                #"asset_tag": "",
                #"description": "",
                #"comments": "",
                replicate_components = False,
                adopt_components= True,
                #"tags": "",
                #"id": "",
            )`

OR

`nb_device.adopt_components = True`

Anyone have any advice?

**Comment by bmerry** [comment]
I've just run into the same problem. I poked around a bit in the code, and it looks like component adoption is in the code handling form submission from the UI, rather than the underlying data model. So it might not be possible to do through the REST API.

**Reply by Kani999** [reply]
We have encountered exactly the same problem and are unable to set `adopt_components` via API calls. 

It seems that `adopt_components` is not defined in `ModuleSerializer`.

**Reply by rusinma2** [reply]
Hello, I am facing the same issue. Is there any solution how to send _adopt_components_ via REST API calls ?

**Reply by Andres1357** [reply]
By coincidence, I also ran into this today. Scoping out how to sync in module data for devices that are already built in NetBox with interfaces. Importing module types & adding the modules via UI using the `adopt_components` option works, but I need to do this via the API.

Also, I do wish `adopt_components` had some capability to overwrite other details about interfaces, specifically, the `type`. My interfaces are typed as "Other" and I would like to populate the type from the module type templates.

**Reply by bmerry** [reply]
The workaround I'm using is not to add interfaces to the module type. Then the module can be added, and afterwards I edit the existing interfaces to reference the module. That would be tedious if done manually but since it's a script it's just a bit of code I had to write once. Naturally the script needs to be aware of which interfaces are intended to go with which module.

For transceivers, this may actually be a more sensible approach, since a transceiver might not have a fixed set of interfaces. A 400Gb SR8 receiver might normally have one associated interface, but it can also be connected to a splitter cable with 2x 200Gb transceivers on the other end, and two interfaces.

**Reply by mytechgnome** [reply]
Has there been any movement on this? I'm running into the same issue. The adopt_components option appears in the GUI, but not in the API. I've tested a module import through the GUI and it works as expected, but when I try via the API it fails.

**Reply by mrmrcoleman** [reply]
Please create an issue for this

**Reply by mytechgnome** [reply]
> Please create an issue for this

I created issue #[21388](https://github.com/netbox-community/netbox/issues/21388)
Not sure if this is a bug or a feature request.
